src/ripple/core/Pg.h

```diff
@@ -0,0 +1,516 @@
+     * existing connection is already present. If connection is not present
+     * or in an errored state, reconnects to the database.
+     */
+    void
+    connect();
+
+    /** Disconnect from postgres. */
+    void
+    disconnect()
+    {
+        conn_.reset();
+    }
+
+    /** Execute postgres query.
+     *
+     * If parameters are included, then the command should contain only a
+     * single SQL statement. If no parameters, then multiple SQL statements
+     * delimited by semi-colons can be processed. The response is from
+     * the last command executed.
+     *
+     * @param command postgres API command string.
+     * @param nParams postgres API number of parameters.
+     * @param values postgres API array of parameter.
+     * @return Query result object.
+     */
+    PgResult
+    query(char const* command, std::size_t nParams, char const* const* values);
+
+    /** Execute postgres query with no parameters.
+     *
+     * @param command Query string.
+     * @return Query result object;
+     */
+    PgResult
+    query(char const* command)
+    {
+        return query(command, 0, nullptr);
+    }
+
+    /** Execute postgres query with parameters.
+     *
+     * @param dbParams Database command and parameter values.
+     * @return Query result object.
+     */
+    PgResult
+    query(pg_params const& dbParams);
+
+    /** Insert multiple records into a table using Postgres' bulk COPY.
+     *
+     * Throws upon error.
+     *
+     * @param table Name of table for import.
+     * @param records Records in the COPY IN format.
+     */
+    void
+    bulkInsert(char const* table, std::string const& records);
+
+public:
+    /** Constructor for Pg class.
+     *
+     * @param config Config parameters.
+     * @param j Logger object.
+     * @param stop Reference to connection pool's stop flag.
+     */
+    Pg(PgConfig const& config, beast::Journal const j, bool& stop)
+        : config_(config), j_(j), stop_(stop)
+    {
+    }
+};
+
+//-----------------------------------------------------------------------------
+
+/** Database connection pool.
+ *
+ * Allow re-use of postgres connections. Postgres connections are created
+ * as needed until configurable limit is reached. After use, each connection
+ * is placed in a container ordered by time of use. Each request for
+ * a connection grabs the most recently used connection from the container.
+ * If none are available, a new connection is used (up to configured limit).
+ * Idle connections are destroyed periodically after configurable
+ * timeout duration.
+ *
+ * This should be stored as a shared pointer so PgQuery objects can safely
+ * outlive it.
+ */
+class PgPool : public Stoppable
+{
+    friend class PgQuery;
+
+    using clock_type = std::chrono::steady_clock;
+
+    PgConfig config_;
+    beast::Journal const j_;
+    std::mutex mutex_;
+    std::condition_variable cond_;
+    std::size_t connections_{};
+    bool stop_{false};
```

Comment:
I think this should be an atomic bool. This variable is read in `checkout`, which could be called from different threads concurrently.

Reply:
In checkout (and each of its other users), it's locked by mutex_ already.

Reply:
Oh you're right. I had just seen the unprotected access at the top of `checkout()`, but it's checked under the mutex later down in that function, so I think it's fine to leave as a normal bool.

Reply:
I'm concerned that this `bool` is read by `Pg::query`, seemingly outside the protection of the mutex.  `Pg::stop_` is a reference to `PgPool::stop_`.

Reply:
Fixed. Good catch.